File: backend/mesh_server.py
import asyncio
import json
import os
import socket
import sys
import time
from typing import Dict, Any
import websockets

# Ensure UTF-8 output on Windows terminal
if hasattr(sys.stdout, 'reconfigure'):
    sys.stdout.reconfigure(encoding='utf-8', errors='replace')
if hasattr(sys.stderr, 'reconfigure'):
    sys.stderr.reconfigure(encoding='utf-8', errors='replace')

PORT = 8765
HISTORY_FILE = os.path.join(os.path.dirname(__file__), 'data', 'mesh_history.json')

connected_clients: Dict[Any, Dict[str, Any]] = {}
sos_history: Dict[str, Dict[str, Any]] = {}

def get_local_ip() -> str:
    try:
        s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        s.connect(('8.8.8.8', 80))
        ip = s.getsockname()[0]
        s.close()
        return ip
    except Exception:
        try:
            return socket.gethostbyname(socket.gethostname())
        except Exception:
            return '127.0.0.1'

def load_history():
    global sos_history
    if os.path.exists(HISTORY_FILE):
        try:
            with open(HISTORY_FILE, 'r', encoding='utf-8') as f:
                data = json.load(f)
                if isinstance(data, list):
                    sos_history = {item['id']: item for item in data if 'id' in item}
                elif isinstance(data, dict):
                    sos_history = data
            print(f"[Mesh Storage] Loaded {len(sos_history)} cached SOS alert(s) from history.")
        except Exception as e:
            print(f"[Mesh Storage] Failed to load history: {e}")

def save_history():
    try:
        os.makedirs(os.path.dirname(HISTORY_FILE), exist_ok=True)
        with open(HISTORY_FILE, 'w', encoding='utf-8') as f:
            json.dump(list(sos_history.values()), f, indent=2)
    except Exception as e:
        print(f"[Mesh Storage] Failed to persist history: {e}")

def get_peer_list(exclude_ws=None):
    peers = []
    for ws, info in connected_clients.items():
        if ws != exclude_ws:
            peers.append({
                "id": info["deviceId"],
                "name": info["name"],
                "ip": info["ip"],
                "connectedAt": info["connectedAt"]
            })
    return peers

async def broadcast_to_peers(message_dict: dict, sender_ws=None):
    raw_payload = json.dumps(message_dict)
    dead_clients = []
    
    for ws in list(connected_clients.keys()):
        if ws == sender_ws:
            continue
        try:
            await ws.send(raw_payload)
        except websockets.ConnectionClosed:
            dead_clients.append(ws)
        except Exception as e:
            print(f"[Broadcast Error] Failed to send: {e}")
            dead_clients.append(ws)

    for dead_ws in dead_clients:
        await handle_disconnect(dead_ws)

async def handle_disconnect(websocket):
    if websocket in connected_clients:
        info = connected_clients.pop(websocket)
        device_id = info.get("deviceId", "Unknown")
        print(f"[-] Node Disconnected: {device_id} ({info.get('name')}) | Remaining peers: {len(connected_clients)}")
        
        leave_msg = {
            "type": "PEER_LEAVE",
            "deviceId": device_id,
            "peers": get_peer_list()
        }
        await broadcast_to_peers(leave_msg)

async def handle_client(websocket):
    remote_ip = websocket.remote_address[0] if websocket.remote_address else "Unknown"
    
    try:
        async for message in websocket:
            try:
                data = json.loads(message)
            except json.JSONDecodeError:
                print(f"[Warning] Invalid JSON from {remote_ip}: {message[:100]}")
                continue

            msg_type = data.get("type")

            if msg_type == "REGISTER":
                device_id = data.get("deviceId", f"node-{id(websocket)}")
                device_name = data.get("name", f"CrisisMesh-{device_id[-4:]}")
                
                connected_clients[websocket] = {
                    "deviceId": device_id,
                    "name": device_name,
                    "ip": remote_ip,
                    "connectedAt": time.time()
                }

                print(f"[+] Node Registered: {device_id} ({device_name}) from {remote_ip}")
                print(f"    Total Active Mesh Nodes: {len(connected_clients)}")

                welcome_packet = {
                    "type": "REGISTER_ACK",
                    "deviceId": device_id,
                    "peers": get_peer_list(exclude_ws=websocket),
                    "serverTime": time.time()
                }
                await websocket.send(json.dumps(welcome_packet))

                if sos_history:
                    sync_packet = {
                        "type": "SYNC_HISTORY",
                        "reports": list(sos_history.values())
                    }
                    await websocket.send(json.dumps(sync_packet))
                    print(f"    -> Synced {len(sos_history)} historical SOS alert(s) to {device_id}")

                join_announcement = {
                    "type": "PEER_JOIN",
                    "peer": {
                        "id": device_id,
                        "name": device_name,
                        "ip": remote_ip,
                        "connectedAt": time.time()
                    },
                    "peers": get_peer_list()
                }
                await broadcast_to_peers(join_announcement, sender_ws=websocket)

            elif msg_type == "SOS":
                sos_data = data.get("data", {})
                sos_id = sos_data.get("id") or f"SOS-{int(time.time()*1000)}"
                sos_data["id"] = sos_id
                priority = sos_data.get("priority", "UNKNOWN")
                origin = sos_data.get("deviceId", "Unknown")
                desc = sos_data.get("description", "")
                num_people = sos_data.get("numPeople", 1)

                print("\n" + "="*60)
                print(f"[!] [EMERGENCY SOS RECEIVED]")
                print(f"   ID:       {sos_id}")
                print(f"   Priority: {priority}")
                print(f"   From:     {origin}")
                print(f"   People:   {num_people}")
                print(f"   Location: Lat {sos_data.get('lat')}, Lng {sos_data.get('lng')}")
                print(f"   Detail:   {desc}")
                print(f"   Broadcasting to {len(connected_clients) - 1} peer node(s)...")
                print("="*60 + "\n")

                sos_history[sos_id] = sos_data
                save_history()

                mesh_packet = {
                    "id": data.get("id", f"msg-{int(time.time()*1000)}"),
                    "type": "SOS",
                    "data": sos_data,
                    "relayedBy": connected_clients.get(websocket, {}).get("deviceId", "Gateway"),
                    "relayCount": (data.get("relayCount") or 0) + 1,
                    "timestamp": int(time.time() * 1000)
                }

                await broadcast_to_peers(mesh_packet, sender_ws=websocket)

            elif msg_type == "PING":
                await websocket.send(json.dumps({"type": "PONG", "timestamp": int(time.time() * 1000)}))

            elif msg_type == "REQUEST_SYNC":
                sync_packet = {
                    "type": "SYNC_HISTORY",
                    "reports": list(sos_history.values())
                }
                await websocket.send(json.dumps(sync_packet))

    except websockets.ConnectionClosed:
        pass
    except Exception as err:
        print(f"[Error in client loop]: {err}")
    finally:
        await handle_disconnect(websocket)

async def main():
    load_history()
    local_ip = get_local_ip()

    print("\n" + "="*65)
    print("  CRISISMESH - OFFLINE PEER-TO-PEER MESH GATEWAY")
    print("="*65)
    print(f"  Status:             ONLINE & LISTENING")
    print(f"  Local Network IP:   {local_ip}")
    print(f"  Port:               {PORT}")
    print(f"  WebSocket URL:      ws://{local_ip}:{PORT}")
    print("="*65)
    print("  Instructions for mobile phones:")
    print(f"  1. Ensure phones & PC are on the same Wi-Fi or Mobile Hotspot.")
    print(f"  2. In the CrisisMesh app Settings, verify Gateway IP is: {local_ip}")
    print("  3. Whenever a node turns on, it will automatically sync.")
    print("="*65 + "\n")

    # Bind to None for dual-stack support (both IPv4 and IPv6)
    async with websockets.serve(handle_client, None, PORT):
        await asyncio.Future()

if __name__ == "__main__":
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        print("\n[CrisisMesh Gateway] Shutting down.")

File: backend/mesh_server.js
const http = require('http');
const os = require('os');
const fs = require('fs');
const path = require('path');
const WebSocket = require(path.join(__dirname, '..', 'frontend', 'node_modules', 'ws'));

const PORT = 8765;
const HISTORY_FILE = path.join(__dirname, 'data', 'mesh_history.json');

let connectedClients = new Map(); // ws -> { deviceId, name, ip, connectedAt }
let sosHistory = new Map(); // id -> sosReport

function getLocalIp() {
  const interfaces = os.networkInterfaces();
  let fallback = '127.0.0.1';
  for (const name of Object.keys(interfaces)) {
    const isVirtual = name.toLowerCase().includes('virtual') || name.toLowerCase().includes('vethernet');
    for (const iface of interfaces[name]) {
      if (iface.family === 'IPv4' && !iface.internal) {
        if (!iface.address.startsWith('192.168.56.') && !isVirtual) {
          return iface.address; // Real Wi-Fi or LAN
        }
        fallback = iface.address;
      }
    }
  }
  return fallback;
}

function loadHistory() {
  if (fs.existsSync(HISTORY_FILE)) {
    try {
      const raw = fs.readFileSync(HISTORY_FILE, 'utf8');
      const data = JSON.parse(raw);
      if (Array.isArray(data)) {
        data.forEach(item => { if (item.id) sosHistory.set(item.id, item); });
      }
      console.log(`[Mesh Storage] Loaded ${sosHistory.size} cached SOS alerts from history.`);
    } catch (e) {
      console.warn('[Mesh Storage] Could not load history:', e.message);
    }
  }
}

function saveHistory() {
  try {
    const dir = path.dirname(HISTORY_FILE);
    if (!fs.existsSync(dir)) fs.mkdirSync(dir, { recursive: true });
    fs.writeFileSync(HISTORY_FILE, JSON.stringify(Array.from(sosHistory.values()), null, 2));
  } catch (e) {
    console.warn('[Mesh Storage] Could not persist history:', e.message);
  }
}

function getPeerList(excludeWs = null) {
  const peers = [];
  for (const [ws, info] of connectedClients.entries()) {
    if (ws !== excludeWs) {
      peers.push({
        id: info.deviceId,
        name: info.name,
        ip: info.ip,
        connectedAt: info.connectedAt
      });
    }
  }
  return peers;
}

function broadcast(messageObj, senderWs = null) {
  const payload = JSON.stringify(messageObj);
  for (const [ws, info] of connectedClients.entries()) {
    if (ws === senderWs) continue;
    if (ws.readyState === WebSocket.OPEN) {
      try {
        ws.send(payload);
      } catch (err) {
        console.warn(`[Broadcast Error] ${info.deviceId}:`, err.message);
      }
    }
  }
}

function handleDisconnect(ws) {
  if (connectedClients.has(ws)) {
    const info = connectedClients.get(ws);
    connectedClients.delete(ws);
    console.log(`[-] Node Disconnected: ${info.deviceId} (${info.name}) | Active nodes: ${connectedClients.size}`);
    broadcast({
      type: 'PEER_LEAVE',
      deviceId: info.deviceId,
      peers: getPeerList()
    });
  }
}

// HTTP Server for /health and diagnostics
const server = http.createServer((req, res) => {
  res.setHeader('Access-Control-Allow-Origin', '*');
  if (req.url === '/health' || req.url === '/') {
    res.writeHead(200, { 'Content-Type': 'application/json' });
    res.end(JSON.stringify({
      status: 'ONLINE',
      gateway: 'CrisisMesh Relay Server',
      activeNodes: connectedClients.size,
      nodes: getPeerList(),
      cachedAlerts: sosHistory.size
    }, null, 2));
  } else {
    res.writeHead(404);
    res.end();
  }
});

const wss = new WebSocket.Server({ server });

wss.on('connection', (ws, req) => {
  const remoteIp = req.socket.remoteAddress || 'Unknown';

  ws.on('message', (message) => {
    try {
      const data = JSON.parse(message.toString());
      const msgType = data.type;

      if (msgType === 'REGISTER') {
        const deviceId = data.deviceId || `node-${Math.random().toString(36).slice(2, 8)}`;
        const name = data.name || `CrisisMesh-${deviceId.slice(-4)}`;

        connectedClients.set(ws, {
          deviceId,
          name,
          ip: remoteIp,
          connectedAt: Date.now()
        });

        console.log(`[+] Node Registered: ${deviceId} (${name}) from ${remoteIp}`);
        console.log(`    Total Active Mesh Nodes: ${connectedClients.size}`);

        // Send ACK
        ws.send(JSON.stringify({
          type: 'REGISTER_ACK',
          deviceId,
          peers: getPeerList(ws),
          serverTime: Date.now()
        }));

        // Send History
        if (sosHistory.size > 0) {
          ws.send(JSON.stringify({
            type: 'SYNC_HISTORY',
            reports: Array.from(sosHistory.values())
          }));
          console.log(`    -> Synced ${sosHistory.size} historical SOS alert(s) to ${deviceId}`);
        }

        // Broadcast PEER_JOIN
        broadcast({
          type: 'PEER_JOIN',
          peer: { id: deviceId, name, ip: remoteIp, connectedAt: Date.now() },
          peers: getPeerList()
        }, ws);

      } else if (msgType === 'SOS') {
        const sosData = data.data || {};
        const sosId = sosData.id || `SOS-${Date.now()}`;
        sosData.id = sosId;
        const priority = sosData.priority || 'UNKNOWN';
        const origin = sosData.deviceId || 'Unknown';
        const desc = sosData.description || '';
        const people = sosData.numPeople || 1;

        console.log('\n============================================================');
        console.log('🚨 [EMERGENCY SOS BROADCAST]');
        console.log(`   ID:       ${sosId}`);
        console.log(`   Priority: ${priority}`);
        console.log(`   From:     ${origin}`);
        console.log(`   People:   ${people}`);
        console.log(`   GPS:      Lat ${sosData.lat}, Lng ${sosData.lng}`);
        console.log(`   Details:  ${desc}`);
        console.log(`   Fanning out to ${connectedClients.size - 1} peer node(s)...`);
        console.log('============================================================\n');

        sosHistory.set(sosId, sosData);
        saveHistory();

        broadcast({
          id: data.id || `msg-${Date.now()}`,
          type: 'SOS',
          data: sosData,
          relayedBy: connectedClients.get(ws)?.deviceId || 'Gateway',
          relayCount: (data.relayCount || 0) + 1,
          timestamp: Date.now()
        }, ws);

      } else if (msgType === 'RESCUE_PING') {
        console.log('\n============================================================');
        console.log('🚑 [ADMIN RESCUE PING DISPATCHED]');
        console.log(`   From:        ${data.adminName || 'Incident Commander'}`);
        console.log(`   Target Node: ${data.targetDeviceId}`);
        console.log(`   SOS ID:      ${data.sosId}`);
        console.log(`   Message:     ${data.message}`);
        console.log(`   Fanning out to all mesh peers...`);
        console.log('============================================================\n');

        broadcast({
          id: data.id || `ping-${Date.now()}`,
          type: 'RESCUE_PING',
          targetDeviceId: data.targetDeviceId,
          sosId: data.sosId,
          adminName: data.adminName,
          message: data.message,
          relayedBy: connectedClients.get(ws)?.deviceId || 'Gateway',
          relayCount: (data.relayCount || 0) + 1,
          timestamp: Date.now()
        }, ws);

      } else if (msgType === 'PEER_CHAT') {
        const chatData = data.chatPayload || data;
        console.log(`💬 [PEER CHAT] From @${chatData.senderUsername || 'User'} (${chatData.senderId}) -> Target: ${chatData.targetDeviceId || 'ALL'}: "${chatData.text}"`);
        broadcast({
          id: data.id || `chat-${Date.now()}`,
          type: 'PEER_CHAT',
          chatPayload: chatData,
          senderId: chatData.senderId,
          senderUsername: chatData.senderUsername,
          targetDeviceId: chatData.targetDeviceId,
          targetUsername: chatData.targetUsername,
          text: chatData.text,
          relayedBy: connectedClients.get(ws)?.deviceId || 'Gateway',
          relayCount: (data.relayCount || 0) + 1,
          timestamp: chatData.timestamp || Date.now()
        }, ws);

      } else if (msgType === 'PING') {
        ws.send(JSON.stringify({ type: 'PONG', timestamp: Date.now() }));
      } else if (msgType === 'REQUEST_SYNC') {
        ws.send(JSON.stringify({
          type: 'SYNC_HISTORY',
          reports: Array.from(sosHistory.values())
        }));
      }
    } catch (e) {
      console.warn('[Warning] Invalid JSON received:', e.message);
    }
  });

  ws.on('close', () => handleDisconnect(ws));
  ws.on('error', () => handleDisconnect(ws));
});

loadHistory();
const localIp = getLocalIp();

server.listen(PORT, '0.0.0.0', () => {
  console.log('\n=================================================================');
  console.log('  CRISISMESH - OFFLINE PEER-TO-PEER MESH GATEWAY (NODE.JS)');
  console.log('=================================================================');
  console.log('  Status:             ONLINE & LISTENING (Dual Stack)');
  console.log(`  Local Network IP:   ${localIp}`);
  console.log(`  Port:               ${PORT}`);
  console.log(`  WebSocket URL:      ws://${localIp}:${PORT}`);
  console.log(`  Health Check URL:   http://${localIp}:${PORT}/health`);
  console.log('=================================================================');
  console.log('  Testing Connectivity:');
  console.log(`  Open http://${localIp}:${PORT}/health on your phone browser.`);
  console.log('  If it loads, the phone can reach your computer!');
  console.log('=================================================================\n');
});
